listenbrainz_spark/hdfs/utils/__init__.py

```diff
@@ -0,0 +1,274 @@
+    # loading newer listens first.
+    file_names.sort(key=lambda x: int(x.split(".")[0]), reverse=True)
+
+    # all incremental dumps are stored in incremental.parquet. these are the newest
+    # listens. but an incremental dump might not always exist for example at the time
+    # when a full dump has just been imported. so check if incremental dumps are
+    # present, if yes then add those to the start of list
+    if has_incremental:
+        file_names.insert(0, "incremental.parquet")
+
+    return file_names
+
+# 4 ->
+
+
+def get_listens_from_new_dump(start: datetime, end: datetime) -> DataFrame:
+    """ Load listens with listened_at between from_ts and to_ts from HDFS in a spark dataframe.
+        Args:
+            start: minimum time to include a listen in the dataframe
+            end: maximum time to include a listen in the dataframe
+        Returns:
+            dataframe of listens with listened_at between start and end
+    """
+    files = get_listen_files_list()
+
+    # create empty dataframe for merging loaded files into it
+    dfs = listenbrainz_spark.session.createDataFrame([], listens_new_schema)
+
+    for file_name in files:
+        df = read_files_from_HDFS(
+            os.path.join(path.LISTENBRAINZ_NEW_DATA_DIRECTORY, file_name)
+        )
+
+        # check if the currently loaded file has any listens newer than the starting
+        # timestamp. if not stop trying to load more files, because listens are sorted
+        # by listened_at in ascending order and we are traversing the files in reverse
+        # order. that is we are loading listens from latest to oldest so if the current
+        # file does not have any listens newer than from_ts, the remaining files will
+        # not have those either.
+        df = df.where(f"listened_at >= to_timestamp('{start}')")
+        if df.count() == 0:
+            break
+
+        # cannot merge this condition with the above one because, consider the following case:
+        # we want listens between the time range - 14 days ago to 7 days ago. the latest file
+        # might have listens only from last 4 days. if the conditions were merged, we would
+        # have stopped looking in other files which is wrong. it is quite possible that the 2nd
+        # or some subsequent file has listens older than to_ts but newer than from_ts
+        df = df.where(f"listened_at <= to_timestamp('{end}')")
+
+        dfs = dfs.union(df)
+
+    return dfs
+
+# 5 ->
+
+def save_parquet(df, path, mode='overwrite'):
+    """ Save dataframe as parquet to given path in HDFS.
+        Args:
+            df (dataframe): Dataframe to save.
+            path (str): Path in HDFS to save the dataframe.
+            mode (str): The mode with which to write the paquet.
+    """
+    try:
+        df.write.format('parquet').save(config.HDFS_CLUSTER_URI + path, mode=mode)
+    except Py4JJavaError as err:
+        raise FileNotSavedException(err.java_exception, path)
+
```

Comment:
The functions till here and `read_json` later in the file should not be moved because those do more than just hdfs stuff. Those functions should remain in the original file.